### PR TITLE
RPP Remap - HOST and HIP Tensor support

### PR DESCRIPTION
**Pull request by snehaa8**

### CODE REVIEW

File: include/rppt_tensor_geometric_augmentations.h [hunk 1]
```diff
@@ -482,6 +482,27 @@ RppStatus rppt_slice_host(RppPtr_t srcPtr, RpptGenericDescPtr srcGenericDescPtr,
 RppStatus rppt_slice_gpu(RppPtr_t srcPtr, RpptGenericDescPtr srcGenericDescPtr, RppPtr_t dstPtr, RpptGenericDescPtr dstGenericDescPtr, RpptROI3DPtr roiGenericPtrSrc, RpptRoi3DType roiType, rppHandle_t rppHandle);
 #endif // GPU_SUPPORT
 
+/******************** remap ********************/
```
Comment (r-abishek) [inline include/rppt_tensor_geometric_augmentations.h:485]: Remove this line as per others in this file

File: include/rppt_tensor_geometric_augmentations.h [hunk 2]
```diff
@@ -482,6 +482,27 @@ RppStatus rppt_slice_host(RppPtr_t srcPtr, RpptGenericDescPtr srcGenericDescPtr,
...
+
+// Remap augmentation for a NCHW/NHWC layout tensor
+
+// *param[in] srcPtr source tensor memory
+// *param[in] srcDesc source tensor descriptor
+// *param[out] dstPtr destination tensor memory
```
Comment (r-abishek) [inline include/rppt_tensor_geometric_augmentations.h:491]: Please change docs as per new Doxygen standard formatting. 
Word it to new standard.

**Review by r-abishek — CHANGES_REQUESTED** [review]
@sampath1117 Please do a round of review on this.

File: src/include/cpu/rpp_cpu_common.hpp [hunk 3]
```diff
@@ -4965,11 +4965,11 @@ inline void compute_generic_bilinear_srclocs_and_interpolate(T *srcPtrChannel, R
     Rpp32f weightParams[4], bilinearCoeffs[4];
     Rpp32s srcLoc[4];
     srcLT.y = (Rpp32s) srcY;                                    // Bilinear LT point y value
-    srcLT.y = std::min(srcLT.y, roiLTRB->ltrbROI.rb.y - 1);
```
Comment (sampath1117) [inline src/include/cpu/rpp_cpu_common.hpp:4968]: I feel this change is needed for other augmentations like resize to match with HIP
Please check and revert this change

File: src/include/cpu/rpp_cpu_common.hpp [hunk 4]
```diff
@@ -4989,19 +4989,19 @@ inline void compute_generic_bilinear_srclocs_and_interpolate(T *srcPtrChannel, R
...
-                    (*(srcPtrChannel + srcLoc[3]) * bilinearCoeffs[3])));        // BottomRow R02 Pixel * coeff3
+                    (*(srcPtrChannel + srcLoc[3]) * bilinearCoeffs[3]));        // BottomRow R02 Pixel * coeff3
         srcPtrChannel += srcDescPtr->strides.cStride;
     }
 }
 
 inline void compute_generic_bilinear_srclocs_1c_avx(__m256 &pSrcY, __m256 &pSrcX, RpptBilinearNbhoodLocsVecLen8 &srcLocs, __m256 *pBilinearCoeffs, __m256 &pSrcStrideH, __m256i *pxSrcStridesCHW, __m256 *pRoiLTRB)
 {
     __m256 pWeightParams[4], pSrcBilinearLTyx[4];
-    pSrcBilinearLTyx[0] = _mm256_floor_ps(pSrcY);                               // srcLT->y = (Rpp32s) srcY;
```
Comment (sampath1117) [inline src/include/cpu/rpp_cpu_common.hpp:5003]: Why this change is reverted?

File: src/include/cpu/rpp_cpu_simd.hpp [hunk 5]
```diff
@@ -2865,10 +2865,17 @@ inline void rpp_store24_f32pln3_to_u8pkd3_avx(Rpp8u* dstPtr, __m256* p)
...
+inline void rpp_store8_u8pln1_to_u8pln1_avx(Rpp8u* dstPtr, __m256i &p)
+{
+    __m128i pTemp = _mm256_castsi256_si128(p);
+    rpp_storeu_si64((__m128i *)(dstPtr), pTemp);
+}
+
 inline void rpp_store8_f32pln1_to_u8pln1_avx(Rpp8u* dstPtr, __m256 &p)
 {
     __m256i px1 = _mm256_permute4x64_epi64(_mm256_packus_epi32(_mm256_cvtps_epi32(p), avx_px0), _MM_SHUFFLE(3,1,2,0));
-    _mm256_storeu_si256((__m256i *)(dstPtr), _mm256_packus_epi16(px1, avx_px0));
```
Comment (sampath1117) [inline src/include/cpu/rpp_cpu_simd.hpp:2871]: why is this change reverted?

File: utilities/test_suite/HIP/Tensor_hip.cpp [hunk 6]
```diff
@@ -312,6 +312,13 @@ int main(int argc, char **argv)
     input_second = static_cast<Rpp8u *>(calloc(inputBufferSize, 1));
     output = static_cast<Rpp8u *>(calloc(outputBufferSize, 1));
 
+    Rpp32f *rowRemapTable = (Rpp32f*) calloc(ioBufferSize, sizeof(Rpp32f));
```
Comment (sampath1117) [inline utilities/test_suite/HIP/Tensor_hip.cpp:315]: Please use a static case here
Also would suggest to allocate only for remap case

You can declare the pointer, but allocate it only for remap cases

File: src/modules/cpu/kernel/remap.hpp [hunk 7]
```diff
@@ -0,0 +1,1949 @@
...
+                                     RppLayoutParams layoutParams)
+{
+    RpptROI roiDefault = {0, 0, (Rpp32s)srcDescPtr->w, (Rpp32s)srcDescPtr->h};
+
+    __m128 pSrcChannel = _mm_set1_ps(srcDescPtr->c);
+    __m128 pSrcStride = _mm_set1_ps(srcDescPtr->strides.hStride);
+
+omp_set_dynamic(0);
+#pragma omp parallel for num_threads(dstDescPtr->n)
+    for(int batchCount = 0; batchCount < dstDescPtr->n; batchCount++)
```
Comment (sampath1117) [inline src/modules/cpu/kernel/remap.hpp:25]: Please change this to use numThreads from handle instead of batchsize

File: src/modules/hip/kernel/remap.hpp [hunk 8]
```diff
@@ -0,0 +1,515 @@
...
+                                RpptDescPtr remapTableDescPtr,
+                                RpptInterpolationType interpolationType,
+                                RpptROIPtr roiTensorPtrSrc,
+                                RpptRoiType roiType,
+                                rpp::Handle& handle)
+{
+    if (roiType == RpptRoiType::LTRB)
+        hip_exec_roi_converison_ltrb_to_xywh(roiTensorPtrSrc, handle);
+
+    int localThreads_x = LOCAL_THREADS_X;
```
Comment (sampath1117) [inline src/modules/hip/kernel/remap.hpp:361]: Please remove the usage of localThreads_x,  localThreads_y,  localThreads_z similar to latest kernel changes

**Review by sampath1117 — COMMENTED** [review]
Please address review comments
@snehaa8 
There has been some functions which are removed/changed quite a bit. Please run QA tests for all functions and ensure all of them pass

File: include/rppt_tensor_geometric_augmentations.h (hunk 1 above)
Comment (snehaa8) [reply include/rppt_tensor_geometric_augmentations.h:485]: Sure made updates as per latest standard

File: include/rppt_tensor_geometric_augmentations.h (hunk 2 above)
Comment (snehaa8) [reply include/rppt_tensor_geometric_augmentations.h:491]: Sure made updates as per latest standard

File: src/include/cpu/rpp_cpu_common.hpp (hunk 3 above)
Comment (snehaa8) [reply src/include/cpu/rpp_cpu_common.hpp]: reverted this change, initially was in an assumption that this fixes mismatch between HIP and HOST

File: utilities/test_suite/HIP/Tensor_hip.cpp (hunk 6 above)
Comment (snehaa8) [reply utilities/test_suite/HIP/Tensor_hip.cpp:315]: Made this change for HOST too

File: include/rppt_tensor_geometric_augmentations.h [hunk 9]
```diff
@@ -482,6 +482,54 @@ RppStatus rppt_slice_host(RppPtr_t srcPtr, RpptGenericDescPtr srcGenericDescPtr,
...
+ * - srcPtr depth ranges - Rpp8u (0 to 255), Rpp16f (0 to 1), Rpp32f (0 to 1), Rpp8s (-128 to 127).
+ * - dstPtr depth ranges - Will be same depth as srcPtr.
+ * \image html img150x150.jpg Sample Input
+ * \image html geometric_augmentations_remap_img150x150.jpg Sample Output
+ * \param [in] srcPtr source tensor in HIP memory
+ * \param [in] srcDescPtr source tensor descriptor (Restrictions - numDims = 4, offsetInBytes >= 0, dataType = U8/F16/F32/I8, layout = NCHW/NHWC, c = 1/3)
+ * \param [out] dstPtr destination tensor in HIP memory
+ * \param [in] dstDescPtr destination tensor descriptor (Restrictions - numDims = 4, offsetInBytes >= 0, dataType = U8/F16/F32/I8, layout = NCHW/NHWC, c = same as that of srcDescPtr)
+ * \param[in] rowRemapTable Rpp32u row numbers for every pixel in the input batch of images
+ * \param[in] colRemapTable Rpp32u column numbers for every pixel in the input batch of images
```
Comment (r-abishek) [inline include/rppt_tensor_geometric_augmentations.h:520]: Pls modify to add a little more one-liner explanation to show what RPP expects - like we say "2D tensor of size batchSize * 4" for the roiTensorSrc argument.

And also specify "in HIP memory" (or HOST memory) for this, since we wont use pinned.

File: src/include/cpu/rpp_cpu_common.hpp [hunk 10]
```diff
@@ -5072,7 +5072,9 @@ inline void compute_generic_bilinear_srclocs_3c_avx(__m256 &pSrcY, __m256 &pSrcX
 template <typename T>
 inline void compute_generic_bilinear_interpolation_pkd3_to_pln3(Rpp32f srcY, Rpp32f srcX, RpptROI *roiLTRB, T *dstPtrTempR, T *dstPtrTempG, T *dstPtrTempB, T *srcPtrChannel, RpptDescPtr srcDescPtr)
 {
-    if ((srcX < roiLTRB->ltrbROI.lt.x) || (srcY < roiLTRB->ltrbROI.lt.y) || (srcX > roiLTRB->ltrbROI.rb.x) || (srcY > roiLTRB->ltrbROI.rb.y))
+    Rpp32s srcXFloor = std::floor(srcX);
+    Rpp32s srcYFloor = std::floor(srcY);
+    if ((srcXFloor < roiLTRB->ltrbROI.lt.x) || (srcYFloor < roiLTRB->ltrbROI.lt.y) || (srcXFloor > roiLTRB->ltrbROI.rb.x) || (srcYFloor > roiLTRB->ltrbROI.rb.y))
```
Comment (r-abishek) [inline src/include/cpu/rpp_cpu_common.hpp:5077]: All changes to bilinear interpolation helpers have been verified with rotate, warp_affine, resize, resize_crop_mirror, RMN etc?

File: src/modules/cpu/kernel/remap.hpp [hunk 11]
```diff
@@ -0,0 +1,1962 @@
+#include "rppdefs.h"
```
Comment (r-abishek) [inline src/modules/cpu/kernel/remap.hpp:1]: Add the license on this file.

File: src/include/cpu/rpp_cpu_common.hpp [hunk 12]
```diff
@@ -5996,4 +5998,21 @@ inline void compute_sum_24_host(__m256d *p, __m256d *pSumR, __m256d *pSumG, __m2
     pSumB[0] = _mm256_add_pd(_mm256_add_pd(p[4], p[5]), pSumB[0]); //add 8B values and bring it down to 4
 }
 
+inline void compute_remap_src_loc_sse(Rpp32f *rowRemapTablePtr, Rpp32f *colRemapTablePtr, Rpp32s *locArray, __m128 &pStride, __m128 &pWidthLimit, __m128 &pHeightLimit, const __m128 &pChannel = xmm_p1)
+{
+    __m128 pRowRemapVal = _mm_loadu_ps(rowRemapTablePtr);
+    pRowRemapVal = _mm_max_ps(_mm_min_ps(pRowRemapVal, pHeightLimit), xmm_p0);
+    __m128 pColRemapVal = _mm_loadu_ps(colRemapTablePtr);
+    pColRemapVal = _mm_max_ps(_mm_min_ps(pColRemapVal, pWidthLimit), xmm_p0);
```
Comment (r-abishek) [inline src/include/cpu/rpp_cpu_common.hpp:6006]: This function seems easily convertible to AVX2. Is there an AVX2 version?

**Review by r-abishek — CHANGES_REQUESTED** [review]

File: src/include/cpu/rpp_cpu_common.hpp (hunk 12 above)
Comment (snehaa8) [reply src/include/cpu/rpp_cpu_common.hpp:6006]: Converting this is simple but its inside vector loop and nearest neighbor implementation is in SSE. Converting nn to AVX will need extra cycles as it will need cross lane shuffling.

File: src/include/cpu/rpp_cpu_simd.hpp (hunk 5 above)
Comment (snehaa8) [reply src/include/cpu/rpp_cpu_simd.hpp:2894]: on rechecking understood that this change is essential, confirmed no other existing kernel fails due to this change

File: utilities/test_suite/HOST/Tensor_host.cpp [hunk 13]
```diff
@@ -862,6 +869,66 @@ int main(int argc, char **argv)
...
+                                rowRemapTableTempRow++;
+                                colRemapTableTempRow++;
+                            }
+                        }
+                    }
+
+                    startWallTime = omp_get_wtime();
+                    startCpuTime = clock();
+                    if (inputBitDepth == 0 || inputBitDepth == 1 || inputBitDepth == 2 || inputBitDepth == 5)
+                        rppt_remap_host(input, srcDescPtr, output, dstDescPtr, rowRemapTable, colRemapTable, tableDescPtr, interpolationType, roiTensorPtrSrc, roiTypeSrc, handle);
```
[the remark this answers is not in the document]
Comment (snehaa8) [reply utilities/test_suite/HOST/Tensor_host.cpp:889]: Done

File: utilities/test_suite/README.md [hunk 14]
```diff
@@ -127,6 +127,10 @@ python runTests.py --case_start 0 --case_end 89 --test_type 0 --qa_mode 1 --batc
 ``` python
 python runTests.py --case_list 21 36 63 --test_type 1 --qa_mode 1 --batch_size 8 --num_runs 100
 ```
+-   QA mode (Performance tests) - Tolerance based PASS/FAIL tests for RPP HIP/HOST functionalities checking achieved improvement in performance percentage over BatchPD versions after comparison to a threshold percentage of improvement
+``` python
+python runTests.py --case_list 21 36 63 --test_type 1 --qa_mode 1 --batch_size 8 --num_runs 100
+```
```
Comment (r-abishek) [inline utilities/test_suite/README.md:133]: Line 130-133 are repeated twice, please selfreview and remove

File: utilities/test_suite/HOST/Tensor_host.cpp [hunk 15]
```diff
@@ -944,6 +951,29 @@ int main(int argc, char **argv)
 
                     break;
                 }
+                case 79:
+                {
+                    testCaseName = "remap";
+
+                    if (interpolationType != RpptInterpolationType::NEAREST_NEIGHBOR && interpolationType != RpptInterpolationType::BILINEAR)
```
Comment (r-abishek) [inline utilities/test_suite/HOST/Tensor_host.cpp:958]: We should be checking this inside the API and return unsupported, not in the test suite

File: src/modules/rppt_tensor_geometric_augmentations.cpp [hunk 16]
```diff
@@ -1149,6 +1149,147 @@ RppStatus rppt_flip_voxel_host(RppPtr_t srcPtr,
...
+                          RpptDescPtr tableDescPtr,
+                          RpptInterpolationType interpolationType,
+                          RpptROIPtr roiTensorPtrSrc,
+                          RpptRoiType roiType,
+                          rppHandle_t rppHandle)
+{
+    RppLayoutParams layoutParams = get_layout_params(srcDescPtr->layout, srcDescPtr->c);
+
+    if (interpolationType != RpptInterpolationType::NEAREST_NEIGHBOR && interpolationType != RpptInterpolationType::BILINEAR)
+        return RPP_ERROR_NOT_IMPLEMENTED;
```
Comment (r-abishek) [inline src/modules/rppt_tensor_geometric_augmentations.cpp:1169]: This check is present on the host, but not in hip in the API below

File: include/rppt_tensor_geometric_augmentations.h [hunk 17]
```diff
@@ -580,6 +580,56 @@ RppStatus rppt_flip_voxel_host(RppPtr_t srcPtr, RpptGenericDescPtr srcGenericDes
...
+ * \image html img150x150.jpg Sample Input
+ * \image html geometric_augmentations_remap_img150x150.jpg Sample Output
+ * \param [in] srcPtr source tensor in HIP memory
+ * \param [in] srcDescPtr source tensor descriptor (Restrictions - numDims = 4, offsetInBytes >= 0, dataType = U8/F16/F32/I8, layout = NCHW/NHWC, c = 1/3)
+ * \param [out] dstPtr destination tensor in HIP memory
+ * \param [in] dstDescPtr destination tensor descriptor (Restrictions - numDims = 4, offsetInBytes >= 0, dataType = U8/F16/F32/I8, layout = NCHW/NHWC, c = same as that of srcDescPtr)
+ * \param[in] rowRemapTable Rpp32u row numbers in HIP memory for every pixel in the input batch of images (1D tensor of size width * height * batchSize)
+ * \param[in] colRemapTable Rpp32u column numbers in HIP memory for every pixel in the input batch of images (1D tensor of size width * height * batchSize)
+ * \param [in] tableDescPtr table tensor descriptor (Restrictions - numDims = 4, offsetInBytes >= 0, dataType = F32, layout = NHWC, c = 1)
+ * \param [in] interpolationType Interpolation type used in <tt> \ref RpptInterpolationType </tt>
```
Comment (r-abishek) [inline include/rppt_tensor_geometric_augmentations.h:621]: Says any interpolation type in RpptInterpolationType, but that's not true. Please specify restrictions on interpolation type somewhere.

**Review by r-abishek — CHANGES_REQUESTED** [review]

File: src/modules/rppt_tensor_geometric_augmentations.cpp (hunk 16 above)
Comment (snehaa8) [reply src/modules/rppt_tensor_geometric_augmentations.cpp:1169]: Added it

File: include/rppt_tensor_geometric_augmentations.h (hunk 17 above)
Comment (snehaa8) [reply include/rppt_tensor_geometric_augmentations.h:621]: Added restrictions

File: utilities/test_suite/HOST/Tensor_host.cpp (hunk 15 above)
Comment (snehaa8) [reply utilities/test_suite/HOST/Tensor_host.cpp:958]: This was added for rotate so added the same here too.

**Comment by snehaa8** [comment]
Addressed comments

**Review by r-abishek — APPROVED** [review]
